apps/gradient_autoscheduler/GradientAutoscheduler.cpp

```diff
@@ -0,0 +1,1008 @@
+            target.has_feature(Halide::Target::AVX512_Skylake) ||
+            target.has_feature(Halide::Target::AVX512_Cannonlake))) {
+        // AVX512F is on all AVX512 architectures
+        return 64 / data_size;
+    } else {
+        return 32 / data_size;
+    }
+}
+
+template <typename FuncOrStage>
+void parallelize_vars_and_rvars_gpu(
+        const MachineParams &params,
+        FuncOrStage func_or_stage,
+        bool is_pure_def,
+        const std::vector<Var> &vars,
+        const std::vector<int> &var_bounds,
+        const std::vector<RVar> &rvars,
+        const std::vector<int> &rvar_bounds,
+        TailStrategy tail,
+        std::ostringstream &schedule_source) {
+    // Find the first variable that has bounds larger or equal than 64,
+    // this is our GPU thread.
+    int split_size = 64;
+    std::vector<Var> gpu_blocks;
+    Var gpu_threads("");
+    int gpu_thread_dim = -1;
+    for (int i = 0; i < (int)vars.size(); i++) {
+        if (gpu_threads.name().empty() && var_bounds[i] >= split_size) {
+            gpu_thread_dim = i;
+            Var outer, inner;
+            func_or_stage.split(vars[i],
+                                outer,
+                                inner,
+                                split_size,
+                                tail);
+            schedule_source << "    .split(" <<
+                vars[i].name() << "," <<
+                outer.name() << "," <<
+                inner.name() << "," <<
+                split_size << "," <<
+                tail << ")\n";
+            gpu_blocks.push_back(outer);
+            gpu_threads = inner;
+        } else {
+            gpu_blocks.push_back(vars[i]);
+        }
+    }
+
+    std::vector<RVar> serial_rvars;
+    std::vector<RVar> r_gpu_blocks;
+    RVar r_gpu_threads("");
+    if (!gpu_threads.name().empty()) {
+        // If we can't find any GPU threads, parallelize RVars to find more parallelism
+        for (int i = 0; i < (int)rvars.size(); i++) {
+            if (!r_gpu_threads.name().empty() && rvar_bounds[i] > split_size) {
+                RVar outer, inner;
+                func_or_stage.split(rvars[i],
+                                    outer,
+                                    inner,
+                                    split_size,
+                                    tail);
+                schedule_source << "    .split(" <<
+                    rvars[i].name() << "," <<
+                    outer.name() << "," <<
+                    inner.name() << "," <<
+                    split_size << "," <<
+                    tail << ")\n";
+                r_gpu_blocks.push_back(outer);
+                r_gpu_threads = inner;
+            } else {
+                r_gpu_blocks.push_back(rvars[i]);
+            }
+        }
+    } else {
+        serial_rvars = rvars;
+    }
+
+    // Fuse all gpu blocks into a single variable
+    Var fused_var("");
+    if (gpu_blocks.size() > 0) {
```

Comment:
here and elsewhere: prefer `!foo.empty()` to `foo.size() > 0`

Reply:
Done